lib/rust/ensogl/component/grid-view/src/lib.rs

```diff
@@ -0,0 +1,508 @@
+            .filter(|loc| !visible_entries.contains_key(loc));
+        uncovered.collect_vec()
+    }
+
+    fn update_after_entries_size_change(&self, properties: Properties) -> Vec<(Row, Col)> {
+        let to_model_request = self.update_entries_visibility(properties);
+        for ((row, col), visible_entry) in &*self.visible_entries.borrow() {
+            set_entry_position(visible_entry, *row, *col, properties.entries_size);
+        }
+        to_model_request
+    }
+
+    fn reset_entries(&self, properties: Properties) -> Vec<(Row, Col)> {
+        let Properties { viewport, entries_size, row_count, col_count } = properties;
+        let mut visible_entries = self.visible_entries.borrow_mut();
+        let mut free_entries = self.free_entries.borrow_mut();
+        let detached = visible_entries.drain().map(|(_, entry)| {
+            entry.unset_parent();
+            entry
+        });
+        free_entries.extend(detached);
+        all_visible_locations(&viewport, entries_size, row_count, col_count).collect_vec()
+    }
+
+    fn drop_all_entries(&self, properties: Properties) -> Vec<(Row, Col)> {
+        let to_model_request = self.reset_entries(properties);
+        self.free_entries.borrow_mut().clear();
+        to_model_request
+    }
+}
+
+impl<E: Entry> Model<E, E::Params> {
+    fn update_entry(
+        &self,
+        row: Row,
+        col: Col,
+        model: E::Model,
+        entry_size: Vector2,
+        text_layer: &Option<WeakLayer>,
+    ) {
+        use std::collections::hash_map::Entry::*;
+        let mut visible_entries = self.visible_entries.borrow_mut();
+        let mut free_entries = self.free_entries.borrow_mut();
+        let create_new_entry = || {
+            let text_layer = text_layer.as_ref().and_then(|l| l.upgrade());
+            self.entry_creation_ctx.create_entry(&text_layer)
+        };
+        let entry = match visible_entries.entry((row, col)) {
+            Occupied(entry) => entry.into_mut(),
+            Vacant(lack_of_entry) => {
+                let new_entry = free_entries.pop().unwrap_or_else(create_new_entry);
+                set_entry_position(&new_entry, row, col, entry_size);
+                self.display_object.add_child(&new_entry);
+                lack_of_entry.insert(new_entry)
+            }
+        };
+        entry.frp().set_model(model);
+    }
+}
+
+
+
+// ================
+// === GridView ===
+// ================
+
+/// A template for [`GridView`] structure, where entry parameters and model are separate generic
+/// arguments.
+///
+/// It may be useful when using GridView in parametrized structs, where we want to avoid rewriting
+/// `Entry` bound in each place. Otherwise it's better to use [`GridView`].
+///
+/// Please note that some bounds are still required, as we use [`Widget`] and [`Frp`] nodes.
+#[derive(CloneRef, Debug, Deref, Derivative)]
+#[derivative(Clone(bound = ""))]
+pub struct GridViewTemplate<E: 'static, M: frp::node::Data, P: frp::node::Data> {
```

Comment:
The names "E", "M", and "P" are not obvious. I spend a lot of time trying to understand what they are. Please name them better or document them. (I'd prefer better naming rather then requiring the reader to go trough docs).